R/visualize_embeddings.R

```diff
@@ -143,77 +144,150 @@ keep_tokens <- function(embedding_df, tokens = "[CLS]") {
+#'     do_pca()
+#' }
+do_pca <- function(embedding_df,
+                   project_vectors = embedding_df,
+                   color_field = NULL,
+                   disambiguate_tokens = TRUE) {
+  num_rows <- nrow(embedding_df)
+  num_rows_proj <- nrow(project_vectors)
+  if (num_rows_proj < 3) {
+    stop("At least three vectors are required for a ",
+         "meaningful PCA plot.")
+  }
+  # Use just the indicated vectors to select the PCA projection.
+  proj_mat <- as.matrix(
+    dplyr::select(project_vectors,
+                  dplyr::matches("V[0-9]+"))
+  )
+  pcs <- stats::prcomp(proj_mat,
+                       retx = TRUE, center = TRUE, scale. = TRUE, rank. = 2L)
+
+  if (disambiguate_tokens) {
+    embedding_df <- dplyr::mutate(embedding_df,
+                                  token = paste(token,
+                                                sequence_index,
+                                                token_index,
+                                                sep = "."))
+  }
+  # Keep all the non-vector columns, for possible use in plotting.
+  tok_labels <- dplyr::select(embedding_df,
+                              -dplyr::matches("V[0-9]+"))
+
+  vec_mat <- as.matrix(
+    dplyr::select(embedding_df,
+                  dplyr::matches("V[0-9]+"))
+  )
+  # Instead of doing PCA here, apply the scaling and projection manually.
+  vec_mat <- scale(vec_mat, center = pcs$center, scale = pcs$scale)
+  projected <- vec_mat %*% pcs$rotation
+
+  pca_df <- dplyr::bind_cols(tok_labels, tibble::as_tibble(projected))
+  return(pca_df)
+}
+
+# plot_pca -------------------------------------------------------------
+
+#' Plot PCA
+#'
+#' Given the output of \code{prep_pca}, make the plot object.
+#'
+#' @param pca_df A tbl_df of pca vectors; output from a call to \code{prep_pca}.
+#' @inheritParams display_pca
+#'
+#' @return A ggplot2 plot of the embedding vectors projected onto two principal
+#'   axes.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # assuming something like the following has been run:
+#' # feats <- RBERT::extract_features(...) # See RBERT documentation
+#' # Then:
+#' feats$output %>%
+#'     filter_layer_embeddings(layer_indices = 12L) %>%
+#'     keep_tokens("[CLS]") %>%
+#'     prep_pca() %>%
+#'     plot_pca()
+#' }
+plot_pca <- function(pca_df,
+                     color_field = NULL) {
+  num_rows <- nrow(pca_df)
+
+  class <- rep("a", num_rows)
+  if (!is.null(color_field)) {
+    if (color_field %in% names(pca_df)) {
+      class <- dplyr::pull(
+        dplyr::select(pca_df, dplyr::one_of(color_field))
+      )
+    } else {
+      warning("Column ", color_field, " not found in input table." )
+    }
+  }
+  class <- as.factor(class)
+
+  unique_classes <- unique(class)
+  num_class <- length(unique_classes)
+  getPalette <- grDevices::colorRampPalette(
+    RColorBrewer::brewer.pal(8, "Dark2")
+  )
+
+  pal <- getPalette(num_class)
+  class_colors <- pal
+  names(class_colors) <- unique_classes
```

Comment:
I feel like all the class stuff should be in the if `!is.null(color_field)` part. If they don't specify a color field, we shouldn't add a scale_color_manual, and we shouldn't specify the color aesthetic. Also, spell out "color" as the aesthetic in any case, cuz I didn't even know "col" could be used and would have been confused if I didn't know what this was.

SO, I'd make the plot, then add scale_color_manual with a color aesthetic iff they asked for that. And then you don't have to deal with the theme bit at all, and the plot is more usable as a ggplot2 object to add more stuff onto.

Reply:
good points!

Reply:
Was putting this off because I needed to think about it. Now... I don't know a good way to specify the color aesthetic only sometimes (besides a large `if` block which has most of the plotting code repeated twice). I think this is why I did it the way I did--so that I could always specify color aesthetic in the `ggplot` call.

Do you know a good way to do this?

Reply:
I *think* this will work:
``` ggp <- ggplot2::ggplot(pca_df, ggplot2::aes(x = PC1, y = PC2,
                                              label = token)) +
    ggplot2::geom_text(vjust = 0, nudge_y = 0.5) +
    ggplot2::geom_point()

if (!is.null(color_field)) {
  blah class stuff
  ggp +
    ggplot2::aes(color = class) +
    ggplot2::scale_color_manual(values = class_colors,
                                name = color_field)
}
```

You can definitely add on the scale_color_manual part optionally. You might have to add the extra aes before the geoms, I can't remember if aes inheritance works for aes's applied after that row. You can also always do a couple lines of `+ ifelse(is.null(color_field), NULL, ggplot2 stuff)`.

Reply:
Wow, I think that *does* work! That is excellent. I don't think the documentation for `aes` indicates anywhere that you can just add an `aes` call as a layer. It just says "aes() is passed to either ggplot() or specific layer..."
Anyway, yeah, that will do it.

Reply:
I think it's better now. I put all the class/color stuff into a single block.